src/components/common/Layout/components/Account/components/Notifications/components/xmtp.ts

```diff
@@ -0,0 +1,142 @@
+import { useCallback, useEffect, useState } from "react"
+import { useFetcherWithSign } from "utils/fetcher"
+import { useAccount, useWalletClient } from "wagmi"
+
+export const useSaveXmtpKeys = () => {
+  const fetcherWithSign = useFetcherWithSign()
+  const { data: signer } = useWalletClient()
+
+  const { id } = useUser()
+
+  return useCallback(
+    async () =>
+      Client.getKeys(signer, { env: "production" }).then(async (key) =>
+        fetcherWithSign([
+          id ? `/v2/users/${id}/keys` : undefined,
+          {
+            body: {
+              key: Buffer.from(key).toString("binary"),
+              service: "XMTP",
+            },
+          },
+        ])
+      ),
+    [id, signer]
+  )
+}
+
+export const useDeleteXmtpKeys = () => {
+  const fetcherWithSign = useFetcherWithSign()
+  const { id } = useUser()
+
+  return async () =>
+    fetcherWithSign([
+      id ? `/v2/users/${id}/keys/1` : undefined,
+      { method: "DELETE" },
+    ])
+}
+
+export const useGetXmtpKeys = () => {
+  const fetcherWithSign = useFetcherWithSign()
+  const [data, setData] = useState(null)
+  const [isLoading, setIsLoading] = useState(true)
+  const [error, setError] = useState(null)
+  const { id, error: userError, isLoading: isUserLoading } = useUser()
+
+  useEffect(() => {
+    if (id) {
+      setIsLoading(true)
+      fetcherWithSign([
+        id ? `/v2/users/${id}/keys` : null,
+        {
+          method: "GET",
+          body: {
+            Accept: "application/json",
+            query: { service: "XMTP" },
+          },
+        },
+      ])
+        .then((response) => {
+          setData(response)
+        })
+        .catch(setError)
+        .finally(() => {
+          setIsLoading(true)
+        })
+    }
+  }, [id])
+
+  return {
+    keys: data?.keys,
+    isLoading: isUserLoading || isLoading,
+    error: userError && error,
+  }
+}
+
+export const useXmtpAccessChecking = () => {
+  const [hasXmtpAccess, setHasAccess] = useState<boolean>(false)
+  const [isCheckingXmtpAccess, setIsCheckingAccess] = useState(false)
+  const { canMessageStatic } = useCanMessage()
+  const showErrorToast = useShowErrorToast()
+
+  const { address } = useAccount()
+
+  useEffect(() => {
+    if (address) {
+      setIsCheckingAccess(true)
+      canMessageStatic(address)
+        .then(setHasAccess)
+        .catch((e) => {
+          console.error("XmtpAccessCheckingError", e)
+          showErrorToast("Error happened during checking XMTP access")
+        })
+        .finally(() => setIsCheckingAccess(false))
+    }
+  }, [address, canMessageStatic])
+  return {
+    hasXmtpAccess,
+    isCheckingXmtpAccess,
+  }
+}
+
+export const useSubscribeXmtp = () => {
+  const [isSubscribingXmtp, setIsSubscribingXmtp] = useState<boolean>(false)
+
+  const toast = useToast()
+  const showErrorToast = useShowErrorToast()
+
+  const { data: signer, isLoading } = useWalletClient()
+
+  const subscribeXmtp = useCallback(async () => {
```

Comment:
I wonder if we could use our `useSubmit` hook here? 👀 That would handle the loading state for us.

Another note: let's call it `subscribeToXMTP`, it feels more natural hehe